modelzoo/MMoE/train.py

```diff
@@ -0,0 +1,897 @@
+    'price': 50
+}
+
+'''
+END OF INPUT CONFIG SPECIFICS
+'''
+'''
+MODEL CONFIG SPECIFICS
+'''
+DNN_ACTIVATION = tf.nn.relu
+
+L2_REGULARIZATION = 1e-06
+EMBEDDING_REGULARIZATION = 5e-05
+
+EXPERTS_COUNT = 4
+EXPERT_HIDDEN_UNITS = [256, 192, 128, 64]
+EMBEDDING_DIM = 16
+
+#Tower tuple structure (tower name, label name, hidden units)
+TOWERS = [
+    ("ctr", "clk", [256, 192, 128, 64]),
+    ("cvr", "buy", [256, 192, 128, 64])
+]
+'''
+MODEL CONFIG SPECIFICS
+'''
+def l2_regularizer(scale, scope=None):
+  if isinstance(scale, numbers.Integral):
+    raise ValueError(f'Scale cannot be an integer: {scale}')
+  if isinstance(scale, numbers.Real):
+    if scale < 0.:
+      raise ValueError(f'Setting a scale less than 0 on a regularizer: {scale}.')
+    if scale == 0.:
+      return lambda _: None
+
+  def l2(weights):
+    with tf.name_scope(scope, 'l2_regularizer', [weights]) as name:
+      my_scale = tf.convert_to_tensor(scale, dtype=weights.dtype.base_dtype, name='scale')
+      return tf.math.multiply(my_scale, tf.nn.l2_loss(weights), name=name)
+
+  return l2
+
+class MMOE():
+    def __init__(self,
+                 input,
+                 feature_column,
+                 num_experts,
+                 expert_hidden_units,
+                 towers,
+                #  l2_scale,
+                 optimizer_type='adam',
+                 learning_rate=0.1,
+                 bf16=False,
+                 stock_tf=None,
+                 adaptive_emb=False,
+                 input_layer_partitioner=None,
+                 dense_layer_partitioner=None):
+        if not input:
+            raise ValueError("Dataset is not defined.")
+        self._feature = input[0]
+        self._label = input[1]
+
+        self._feature_column = feature_column
+        self._num_experts = num_experts
+        self._expert_hidden_units = expert_hidden_units
+        self._towers = towers
+
+        # self.__l2_regularization = l2_regularizer(l2_scale) if l2_scale else None
+        self._learning_rate = learning_rate
+        self.tf = stock_tf
+        self._bf16 = False if self.tf else bf16
+
+        self.is_training = True
+        self._adaptive_emb = adaptive_emb
+        self._optimizer_type = optimizer_type
+        self._input_layer_partitioner = input_layer_partitioner
+        self._dense_layer_partitioner = dense_layer_partitioner
+
+        self.model = self._create_model()
+        with tf.name_scope('head'):
+            self._create_loss()
+            self._create_optimizer()
+            self._create_metrics()
+    
+    # used to add summary in tensorboard
+    def _add_layer_summary(self, value, tag):
+        tf.summary.scalar('%s/fraction_of_zero_values' % tag,
+                        tf.nn.zero_fraction(value))
+        tf.summary.histogram('%s/activation' % tag, value)
+    
+    def _make_scope(self, name, bf16, part):
+        if(bf16):
+            return tf.variable_scope(name, partitioner=part, reuse=tf.AUTO_REUSE).keep_weights()
```

Comment:
keep_weights(dtype=tf.float32), this makes it clearer what the function does.

Reply:
Changed